### PR TITLE
Instance flattener

### LINKED ISSUE #1101: Implement the Instance Flattener
Described in [ADR 7](https://github.com/informalsystems/quint/blob/main/doc/adr007-flattening.md)

### PR DESCRIPTION
**Pull request by bugarela**
Hello :octocat: 

This introduces the instance flattener described in [ADR 07](https://github.com/informalsystems/quint/blob/main/doc/adr007-flattening.md). Closes #1101.

This is not connected to any code for the moment. My next PR will wire this in.

- [X] Tests added for any new code
- [X] Documentation added for any new functionality
- [ ] Entries added to the respective `CHANGELOG.md` for any new functionality
- [ ] Feature table on [`README.md`](../README.md#roadmap) updated for any listed functionality

### CODE REVIEW

File: quint/src/flattening/instanceFlattener.ts [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    const newModuleDecls: QuintDeclaration[] = []
+
+    decl.overrides.forEach(([param, expr]) => {
+      // Find dependencies of `expr` and add those to the new module
+      newModuleDecls.push(...dependentDefinitions(expr, this.lookupTable))
+
+      if (expr.kind === 'name' && expr.name === param.name) {
+        // Special case for instances like `import A(x = x) ...`
+        // In this case, the definition for `x` would have already been added by the dependentDefinitions call above
+        // so there is no need to add a new definition for `x` here. In fact, that would introduce a conflict,
```
Comment (konnov) [inline quint/src/flattening/instanceFlattener.ts:128]:
```suggestion
        // so there is no need to add a new definition for `x` here. In fact, that would introduce a conflict.
```

**Review by konnov — APPROVED** [review]
Looks good!

**Review by thpani — APPROVED** [review]
Amazing, LGTM!

File: quint/src/flattening/instanceFlattener.ts [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    // the name, as well as the qualifier (or the proto name itself, if no qualifier is present)
+    // If the current module is named "myModule":
+    // - The module for `import A.*` will be named `myModule::A`
+    // - The module for `import A as B.*` will be named `myModule::B`
+    const newName = [this.currentModuleName!, decl.qualifiedName ?? protoModule.name].join('::')
+
+    const transformedDecls = newModuleDecls.map(decl => {
+      if (!isDef(decl)) {
+        // Keep imports/exports/instances as is
+        // TODO: further test and explore scenarios where the proto module has imports/exports/instances
```
[the remark this answers is not in the document]
Comment (bugarela) [reply quint/src/flattening/instanceFlattener.ts:149]: Sure! I'll make a general issue to explore corner cases and add this one

**Pull request merged**